### PR TITLE
DPE Move Patch Handling

### PR DESCRIPTION
**Pull request by alexmontAmazon**
image

## What does this PR do?
MAJOR STUFF!
- implemented move patch handling in the DPE!
- implemented sort enabling/disabling in the SortAdapter which now generate optimal Move patch operations
- added a base RowSortAdapter::LessThan that maintains the sorting of the source adapter. This allows RowSortAdapter derived classes to stabilize their sort in the case of ties
- implemented RowSortAdapter::ResortRow now that the DPE supports move operations
- moved recycling/ownership of Handlers out of DPERowWidget. The logic is now in the DPE itself, but the ownership is now tied to the handlerWidget via Qt's property system

LESS MAJOR STUFF!
- fixed PatchOperation::ApplyMove() to work for moves within the same parent
- added Path::Back() convenience function
- fixed error when disabling sort before tree was even generated
- fixed and optimized RowSortAdapter::MapPath()
- added a base class "SortInfoBase" for SortInfoNode to easily search within the sets
- removed as-of-yet unimplemented reverse sort cruft

## How was this PR tested?
added sorting to the DPEDebugViewStandalone project and tested locally

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/SortAdapter.cpp [hunk 1]
```diff
@@ -20,26 +21,31 @@ namespace AZ::DocumentPropertyEditor
...
             if (active)
             {
-                // TODO: This method should generate patches to/from source once custom patching is implemented
-                HandleReset();
-
                 // we can use the currently cached tree, if it still exists. NB it will be cleared by DOM changes
                 if (!m_rootNode)
                 {
-                    // TODO: GenerateFullTree();
+                    GenerateFullTree();
```
Comment (cgalvan) [inline Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/SortAdapter.cpp:31]: Cool :)

**Review by cgalvan — APPROVED** [review]

**Review by amzn-tmryan — APPROVED** [review]
Congrats! This was a doozy.

I see the concept of index vs adapter being referenced in variable names. I don't quite understand the difference after reading the PR though. Is it something that's worth explaining in a short comment somewhere?

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/SortAdapter.cpp [hunk 2]
```diff
@@ -20,26 +21,31 @@ namespace AZ::DocumentPropertyEditor
...
         {
             m_sortActive = active;
+
+            Dom::Patch outgoingPatch;
             if (active)
             {
-                // TODO: This method should generate patches to/from source once custom patching is implemented
-                HandleReset();
-
                 // we can use the currently cached tree, if it still exists. NB it will be cleared by DOM changes
```
[the remark this answers is not in the document]
Comment (alexmontAmazon) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/SortAdapter.cpp:28]: ooh sorry, that's an old-school convention I follow. It stands for "nota bene" or "note well":
https://en.wikipedia.org/wiki/Nota_bene

[the remark this answers is not in the document]
Comment (amzn-tmryan) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/SortAdapter.cpp:28]: No worries! Thanks for the explanation :)

**Pull request merged**